Basic pagination for job-based endpoints (#9397)
fishtown-analytics/docs.getdbt.com: website/docs/docs/dbt-apis/discovery-querying.md

totalCount # Total number of records across all pages
```

The previously described `PageInfo` and `totalCount` pattern applies to the `environment` endpoints, which return results as a connection of `edges` and `nodes`.

#### Job-based endpoint pagination

> Job-based list endpoints also support cursor-based pagination. Use this pattern when you query resources under the [`job`](/docs/dbt-apis/discovery-schema-job) object (or the equivalent legacy top-level queries such as `models(jobId:, runId:)`). These endpoints return a flat list rather than an `edges`/`nodes` connection, so they do not expose a `PageInfo` object.

Reviewer: let's not mention this portion

"(or the equivalent legacy top-level queries such as `models(jobId:, runId:)`)"

There's no traffic to those and they're marked as deprecated because they don't support bigint (required in a multicell world)

> Job-based list endpoints also support cursor-based pagination. Use this pattern when you query resources under the [`job`](/docs/dbt-apis/discovery-schema-job) object (or the equivalent legacy top-level queries such as `models(jobId:, runId:)`). These endpoints return a flat list rather than an `edges`/`nodes` connection, so they do not expose a `PageInfo` object.

Reviewer 2: ```suggestion
Job-based list endpoints also support cursor-based pagination. Use this pattern when you query resources under the [`job`](/docs/dbt-apis/discovery-schema-job) object. These endpoints return a flat list rather than an `edges`/`nodes` connection, so they do not expose a `PageInfo` object.
```

The following job-based list fields accept `first` and `after`:

- `models`
- `sources`
- `seeds`
- `snapshots`
- `tests`
- `macros`
- `metrics`
- `exposures`

Pass `first` to set the page size (capped at 100). Each returned item includes a `paginationCursor` field—an opaque, per-element value you select in your query. Pass the last item's `paginationCursor` as `after` on the next request to fetch the following page.

Unlike environment queries, these job-based endpoints do not return a `PageInfo` object or `hasNextPage` field. You have reached the last page when a page returns fewer rows than `first` (or fewer than the default page size).

> We recommend always specifying `first` to keep response sizes manageable. If you omit `first` and `after`, the API returns all matching rows today. For large jobs, pass `first` and `after` explicitly so your integration does not depend on unpaginated responses.

Reviewer: "If you omit `first` and `after`, the API returns all matching rows today"

Even with the note regarding an upcoming change below, I think we shouldn't say "the API returns all matching rows today". Would it be possible to change it to something like "historically, the API returned all rows"? I don't want to encourage any newer use cases to rely on that behavior.

> We recommend always specifying `first` to keep response sizes manageable. If you omit `first` and `after`, the API returns all matching rows today. For large jobs, pass `first` and `after` explicitly so your integration does not depend on unpaginated responses.

Reviewer 2: ```suggestion
We recommend always specifying `first` to keep response sizes manageable. Historically, omitting `first` and `after` would cause the API to return all matching rows. For large jobs, best practice is to pass `first` and `after` explicitly so your integration does not depend on unpaginated responses.
```

totalCount # Total number of records across all pages
```

> #### Job-based queries

Reviewer 3: Love the attention to detail here, but this H4 is almost three times as long as the H3 it sits under or any other section on this page. Is there any way to condense it or to link out to the API docs for the users to test examples instead of covering it step-by-step here? Not a problem if you feel this level of detail is required to accurately convey the feature

> Job-based list endpoints also support cursor pagination. Use this pattern when you query resources under the [`job`](/docs/dbt-apis/discovery-schema-job) object (or the equivalent legacy top-level queries such as `models(jobId:, runId:)`).

Reviewer 3: ```suggestion
Job-based list endpoints also support cursor-based pagination. Use this pattern when you query resources under the [`job`](/docs/dbt-apis/discovery-schema-job) object (or the equivalent legacy top-level queries such as `models(jobId:, runId:)`).
```

The following job-based list fields accept `first` and `after`:

- `models`
- `sources`
- `seeds`
- `snapshots`
- `tests`
- `macros`
- `metrics`
- `exposures`

Pass `first` to set the page size (capped at 100). Each returned item includes a `paginationCursor` field. Pass that value as `after` on the next request to fetch the following page.

Unlike environment queries, these job-based endpoints do not return a `PageInfo` object or `hasNextPage` field. When a page contains fewer rows than `first`, you have reached the last page.

> If you omit `first` and `after`, the API still returns all matching rows today. In a future update, unpaginated requests will return at most 100 rows. That change will roll out after advance notice. Until then, pass `first` and `after` explicitly when you query large jobs.

Reviewer 3: Maybe this should be a callout. We can alter it as the dates are defined. Makes it easier to identify when it's time to remove as well

   ```text
   /deploy/<account>/projects/<project>/runs/<runId>
   ```

   If you omit `runId` in the query, the API uses the job's latest run.

7. Run the query with the [GraphQL explorer](#run-queries-with-the-graphql-explorer) or [HTTP requests](#run-queries-using-http-requests). You can also look up job and run IDs with the [Admin API](/docs/dbt-apis/admin-api).

**First page**

Query:

```graphql
query JobModelsPage($jobId: BigInt!, $runId: BigInt, $first: Int!, $after: String) {
  job(id: $jobId, runId: $runId) {
    models(first: $first, after: $after) {
      uniqueId
      paginationCursor
    }
  }
}
```

Variables (replace the IDs with your job and run):

```json
{
  "jobId": 12345,
  "runId": 67890,
  "first": 10,
  "after": null
}
```

Example response (truncated):

```json
{
  "data": {
    "job": {
      "models": [
        {
          "uniqueId": "model.my_project.stg_orders",
          "paginationCursor": "Y3Vyc29yOm1vZGVsLm15X3Byb2plY3Quc3RnX29yZGVycw=="
        },
        {
          "uniqueId": "model.my_project.dim_customers",
          "paginationCursor": "Y3Vyc29yOm1vZGVsLm15X3Byb2plY3QuZGltX2N1c3RvbWVycw=="
        }
      ]
    }
  }
}
```

**Next page**

> Set `after` to the `paginationCursor` from the **last** row of the previous page. The cursor is an opaque encoded string, not the `uniqueId`. Keep `first` the same unless you want a different page size.

Reviewer 3: ```suggestion
Set `after` to the `paginationCursor` from the _last_ row of the previous page. The cursor is an opaque encoded string, not the `uniqueId`. Keep `first` the same unless you want a different page size.
```
